NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1060,3 +1097,133 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length) const
+            if (pPage == nullptr)
+            {
+                DPrintf(0, "[%s] no available shared pages !\n", __FUNCTION__);
+                return false;
+            }
+            else
+            {
+                m_Context->SendCopyPagesInUsed.Push(pPage);
+                m_UsedPages[i] = pPage;
+                m_UsedPagesCount++;
+                PVOID pVirtualAddress = pPage->GetVA();
+                ULONG uCopySize = (i == uPages - 1) ? (uDataLength - uDataOffset) : PAGE_SIZE;
+                ULONG uCopied = Copy(pVirtualAddress, uCopySize, uDataOffset);
+                if (uCopied != uCopySize)
+                {
+                    DPrintf(0, "[%s] copy failed! expected %lu, copied %lu bytes\n", __FUNCTION__, uCopySize, uCopied);
+                    return false;
+                }
+                m_CopySGL->Elements[i].Address = pPage->GetPA();
+                m_CopySGL->Elements[i].Length  = uCopySize;
+                uDataOffset += uCopySize;  // Adjust offset, last page might not need adjustment but does not harm
+            }
+        }
+        return true;
+    }
+#endif
+
+
+#if MAX_FRAGMENTS_IN_ONE_NB
+    bool CNB::MapCopyDataToVirtioSGL(CTXDescriptor &Descriptor, ULONG Offset) const
+    {
+        for (ULONG i = 0; i < m_CopySGL->NumberOfElements; i++)
+        {
+            if (Offset < m_CopySGL->Elements[i].Length)
+            {
+                PHYSICAL_ADDRESS PA;
+                PA.QuadPart = m_CopySGL->Elements[i].Address.QuadPart + Offset;
+
+                if (!Descriptor.AddDataChunk(PA, m_CopySGL->Elements[i].Length - Offset))
+                {
+                    return false;
+                }
+                Offset = 0;
+            }
+            else
+            {
+                Offset -= m_CopySGL->Elements[i].Length;
+            }
+        }
+
+        return true;
+    }
+#endif
+
+#if MAX_FRAGMENTS_IN_ONE_NB
+    ULONG CNB::Copy(PVOID Dst, ULONG Length, ULONG DataOffset) const
+    {
+        //find the MDL where the offset position, DataOffset, is located.
+        ULONG Offset = 0;
+        PMDL DataOffsetMDL = nullptr;
+        for (DataOffsetMDL = NET_BUFFER_CURRENT_MDL(m_NB);; DataOffsetMDL = DataOffsetMDL->Next)
+        {
+            Offset += MmGetMdlByteCount(DataOffsetMDL);
+            if (Offset > DataOffset + NET_BUFFER_CURRENT_MDL_OFFSET(m_NB) || DataOffsetMDL->Next == nullptr)
+            {
+                break;
+            }
+        }
+        ULONG CurrOffset = MmGetMdlByteCount(DataOffsetMDL) - (Offset - DataOffset - NET_BUFFER_CURRENT_MDL_OFFSET(m_NB));
+        ULONG Copied = 0;
+
+        Length = min(Length, NET_BUFFER_DATA_LENGTH(m_NB) - DataOffset);
+             
+        for (PMDL CurrMDL = DataOffsetMDL;
+            CurrMDL != nullptr && Copied < Length;
+            CurrMDL = CurrMDL->Next)
+        {
+            ULONG CurrLen;
+            PVOID CurrAddr = nullptr;
+
+    #if NDIS_SUPPORT_NDIS620
```

Comment:
In today's repository we're always with NDIS650 or above

Reply:
@ybendito  Okay, I will delete this #if NDIS_SUPPORT_NDIS620. 
Additionally, I wanted to ask, I noticed that there are many  #if NDIS_SUPPORT_NDIS620 in other parts of the code; are they also outdated and need to be removed?

Reply:
@zjmletang Probably we can remove unneeded ifdefs also in the existing code, but not in this PR. With new code we definitely do not need to take care of older OSes.